.pre-commit-config.yaml

```diff
@@ -0,0 +1,55 @@
+repos:
+- repo: https://github.com/pre-commit/pre-commit-hooks
+  rev: v5.0.0
+  hooks:
+    - id: trailing-whitespace
+    - id: check-ast
+    - id: debug-statements
+    - id: end-of-file-fixer
+    # - id: check-docstring-first
+    - id: check-added-large-files
+      exclude_types: [yaml]
+
+# - repo: https://github.com/keewis/blackdoc
+#   rev: v0.3.9
+#   hooks:
+#     - id: blackdoc
+
+- repo: https://github.com/codespell-project/codespell
+  rev: v2.4.1
+  hooks:
+    - id: codespell
+      exclude: >
+          (?x)^(
+              .*\.ipynb
+          )$
+      args:
+        - --ignore-words-list=astroid,fo
+
+- repo: https://github.com/astral-sh/ruff-pre-commit
+  rev: v0.11.12
+  hooks:
+    - id: ruff
+      args: ["--fix", "--show-fixes"]
+    # - id: ruff-format
```

Comment:
If folks want to adopt ruff format in lieu of black, we should do that ASAP to avoid a lot of churn in the code base as the code grows in size.

Reply:
see below -- yes, let's use ruff.

Reply:
Got it -- I added some ruff format config, and ran it, in the reformat branch.

It seems to have done a fine job, with not a lot of disruption.

Can you merge that in to this pre-commit branch and see how it goes?